shm_wrapper_aux/shm_wrapper_aux.c

```diff
@@ -0,0 +1,469 @@
+	}
+	printf("\n");
+	fflush(stdout);
+	
+	//release rd_sem
+	if (sem_post(gp_sem) == -1) {
+		error("sem_post: gamepad_mutex (in print)");
+	}
+}
+
+// ************************************ PUBLIC WRAPPER FUNCTIONS ****************************************** //
+
+/*
+Call this function from every process that wants to use the auxiliary shared memory wrapper
+Should be called before any other action happens
+The network handler process is responsible for initializing the shared memory blocks
+	- process: one of DEV_HANDLER, EXECUTOR, NET_HANDLER to specify which process this function is
+		being called from
+No return value.
+*/
+void shm_aux_init (process_t process)
+{
+	int fd_shm; //file descriptor of the memory-mapped shared memory
+	
+	if (process == NET_HANDLER) {
+		
+		//mutual exclusion semaphore, gp_sem with initial value = 0
+		if ((gp_sem = sem_open(GP_MUTEX_NAME, O_CREAT, 0660, 0)) == SEM_FAILED) {
+			error("sem_open: gamepad_mutex@net_handler");
+		}
+		
+		//mutual exclusion semaphore, rd_sem with initial value = 1
+		if ((rd_sem = sem_open(RD_MUTEX_NAME, O_CREAT, 0660, 1)) == SEM_FAILED) {
+			error("sem_open: robot_desc_mutex@net_handler");
+		}
+		
+		//create gamepad shm block
+		if ((fd_shm = shm_open(GPAD_SHM_NAME, O_RDWR | O_CREAT, 0660)) == -1) {
+			error("shm_open gamepad_shm: @net_handler");
+		}
+		if (ftruncate(fd_shm, sizeof(gamepad_t)) == -1) {
+			error("ftruncate gamepad_shm: @net_handler");
+		}
+		if ((gp_ptr = mmap(NULL, sizeof(gamepad_t), PROT_READ | PROT_WRITE, MAP_SHARED, fd_shm, 0)) == MAP_FAILED) {
+			error("mmap gamepad_shm: @net_handler");
+		}
+		if (close(fd_shm) == -1) {
+			error("close gamepad_shm: @net_handler");
+		}
+		
+		//create robot description shm block
+		if ((fd_shm = shm_open(ROBOT_DESC_SHM_NAME, O_RDWR | O_CREAT, 0660)) == -1) {
+			error("shm_open robot_desc_shm: @net_handler");
+		}
+		if (ftruncate(fd_shm, sizeof(robot_desc_t)) == -1) {
+			error("ftruncate robot_desc_shm: @net_handler");
+		}
+		if ((rd_ptr = mmap(NULL, sizeof(robot_desc_t), PROT_READ | PROT_WRITE, MAP_SHARED, fd_shm, 0)) == MAP_FAILED) {
+			error("mmap robot_desc_shm: @net_handler");
+		}
+		if (close(fd_shm) == -1) {
+			error("close robot_desc_shm: @net_handler");
+		}
+		
+		//initialize everything
+		gp_ptr->buttons = 0;
+		for (int i = 0; i < 4; i++) {
+			gp_ptr->joysticks[i] = 0.0;
+		}
+		rd_ptr->fields[STATE] = NOMINAL;
+		rd_ptr->fields[RUN_MODE] = IDLE;
+		rd_ptr->fields[DAWN] = DISCONNECTED;
+		rd_ptr->fields[SHEPHERD] = DISCONNECTED;
+		rd_ptr->fields[GAMEPAD] = DISCONNECTED;
+		rd_ptr->fields[TEAM] = BLUE;                //arbitrary
+		rd_ptr->net_to_exe_pending = 0;             //no writes made yet
+		rd_ptr->exe_to_net_pending = 0;             //no writes made yet
+		
+		//initialization complete; set gp_mutex to 1 indicating shm segment available for client(s)
+		if (sem_post(gp_sem) == -1) {
+			error("sem_post: gamepad_mutex@net_handler");
```

Comment:
1. Why is this needed? 2. Why don't you also post rd_sem?

Reply:
1) It's needed because I initialize gp_sem with an initial value of 0 on line 159. Basically it makes sure that all the shared memory is created and initialized before other processes try to attach to it, because other processes will wait to open the gp_sem and that will block onto gp_sem is posted.
2) I don't need to post rd_sem because I initialize it with an initial value of 1 (unlocked) on line 163. I arbitrarily chose gp_sem to be the one that blocks other processes from trying to attach to shared memory before it's created.